src/net/mqtt_interface.rs

```diff
@@ -0,0 +1,235 @@
+            phy,
+            telemetry_topic,
+            default_response_topic: response_topic,
+            network_was_reset: false,
+            subscribed: false,
+        }
+    }
+
+    /// Update the MQTT interface and service the network
+    ///
+    /// # Returns
+    /// An option containing an action that should be completed as a result of network servicing.
+    pub fn update(&mut self) -> Option<Action> {
+        // First, service the network stack to process any inbound and outbound traffic.
+        let sleep = match self
+            .mqtt
+            .borrow_mut()
+            .network_stack
+            .poll(self.clock.current_ms())
+        {
+            Ok(updated) => !updated,
+            Err(err) => {
+                log::info!("Network error: {:?}", err);
+                false
+            }
+        };
+
+        // If the PHY indicates there's no more ethernet link, reset the DHCP server in the network
+        // stack.
+        if self.phy.poll_link() == false {
+            // Only reset the network stack once per link reconnection. This prevents us from
+            // sending an excessive number of DHCP requests.
+            if !self.network_was_reset {
+                self.network_was_reset = true;
+                self.mqtt.borrow_mut().network_stack.handle_link_reset();
+            }
+        } else {
+            self.network_was_reset = false;
+        }
+
+        let mqtt_connected = match self.mqtt.borrow_mut().is_connected() {
+            Ok(connected) => connected,
+            Err(minimq::Error::Network(
+                smoltcp_nal::NetworkError::NoIpAddress,
+            )) => false,
+            Err(minimq::Error::Network(error)) => {
+                log::info!("Unexpected network error: {:?}", error);
+                false
+            }
+            Err(error) => {
+                log::warn!("Unexpected MQTT error: {:?}", error);
+                false
+            }
+        };
+
+        // If we're no longer subscribed to the settings topic, but we are connected to the broker,
+        // resubscribe.
+        if !self.subscribed && mqtt_connected {
+            let mut settings_topic: String<consts::U128> = String::new();
+            write!(&mut settings_topic, "{}/settings/#", self.id.as_str())
+                .unwrap();
+
+            self.mqtt
+                .borrow_mut()
+                .subscribe(&settings_topic, &[])
+                .unwrap();
+            self.subscribed = true;
+        }
+
+        // Handle any MQTT traffic.
+        let mut update = false;
+        match self.mqtt.borrow_mut().poll(
+            |client, topic, message, properties| {
+                let (response, settings_update) =
+                    self.route_message(topic, message, properties);
+                client
+                    .publish(
+                        response.topic,
+                        &response.message,
+                        minimq::QoS::AtMostOnce,
```

Comment:
Let's add a `TODO` that this should be `AtLeastOnce` or even `ExactlyOnce` (typo in minimq).